buildpack/telemetry/appdynamics_telegraf_output.py

```diff
@@ -0,0 +1,126 @@
+"""The module is used in runtime to convert metrics from telegraf for appdynamics Machine Agent"""
+import json
+import logging
+import requests
+from requests.exceptions import ConnectionError
+
+
+METRIC_TREE_BASE_NODE = "Custom Metrics|mxapp"
+AGGR_TYPE = "OBSERVATION"
+
+
+def filter_last(payload):
+
+    last_metrics = {}
+
+    for metric in payload:
+
+        metric_name = metric["metricName"]
+        last_metric = last_metrics.get(metric_name)
+
+        if last_metric:
+            if metric["timestamp"] > last_metric["timestamp"]:
+                last_metrics[metric_name] = metric
+        else:
+            last_metrics[metric_name] = metric
+
+    filtered_metrics = []
+
+    for metric in last_metrics.values():
+        del metric["timestamp"]
+        filtered_metrics.append(metric)
+
+    return filtered_metrics
+
+
+def convert_metric(metric):
+
+    fields = metric.get("fields")
+
+    converted = []
+
+    if fields is None:
+        logging.error("Invalid format of specific metric (telegraf)")
+        return
+
+    for value_name in fields.keys():
+        value = fields[value_name]
+        if value_name == "value":
+            metric_name = metric["name"]
+        else:
+            metric_name = "_".join((metric["name"], value_name))
+
+        metric_path_list = [METRIC_TREE_BASE_NODE, metric_name]
+
+        tags = metric.get("tags")
+        if tags:
+            area = tags.get("area")
+            id = tags.get("id")
+
+            if area:
+                metric_path_list.append(area)
+            if id:
+                metric_path_list.append(id)
+
+        metric_path = "|".join(metric_path_list)
+
+        conv_metric = {
+            "metricName": metric_path,
+            "aggregatorType": AGGR_TYPE,
+            "value": value,
+            "timestamp": metric["timestamp"]
+        }
+
+        converted.append(conv_metric)
+
+    return converted
+
+
+def convert_payload():
+
+    metrics_str = input()
+    metrics_dict = json.loads(metrics_str)
+    metrics_list = metrics_dict.get("metrics")
+
+    with open("/tmp/test.json", 'w') as f:
+        json.dump(metrics_dict, f, indent=4)
+
+    if metrics_list is None:
+        logging.error("Invalid format of metrics json (telegraf)")
+        return
+
+    appdynamics_payload = []
+
+    for metric in metrics_list:
+
+        converted_metrics = convert_metric(metric)
```

Comment:
Do add some documentation here to explain why we need to do the convert and maybe add links to the docs detailing the payload machine agent expects.

Reply:
Done